lighthouse-core/lib/network-request.js

```diff
@@ -202,6 +202,15 @@ class NetworkRequest {
     this.frameId = data.frameId;
   }
 
+  /**
+   * @param {LH.Crdp.Network.ResponseReceivedExtraInfoEvent} data
+   */
+  onResponseReceivedExtraInfo(data) {
```

Comment:
I haven't tested anything, so based purely on the protocol docs, Joey's comment in https://github.com/GoogleChrome/lighthouse/issues/9738#issuecomment-535217004, and his review with caseq in https://chromium-review.googlesource.com/c/chromium/src/+/1639277, it seems like this might not be sufficient to reconcile the extra info.

Two possible issues:
- from `ResponseReceivedExtraInfoEvent`: "responseReceivedExtraInfo may be fired before or after responseReceived". If a `ResponseReceivedEvent` comes in after the extra info event, it would override the updated headers (this matches up with the guard added in that CL and still in the frontend code today)
- it looks like `ResponseReceivedExtraInfoEvent`s can come in _after_ redirects have occurred. `RedirectExtraInfoBuilder` is used in the DT frontend's NetworkManager to match with previous requests when redirects have occurred because they all share a requestId there. We handle that with unique requestIds and `_findRealRequestAndSetSession`, but the extra-info-event ordering would seem(?) to violate our "Once a request has been redirected all future messages referencing that requestId are about the new destination, not the original" invariant on that method.

cc @patrickhulce if you have any network request insights

Reply:
> updated headers

Why would the headers be updated / differ between these events?

> We handle that with unique requestIds and _findRealRequestAndSetSession, but the extra-info-event ordering would seem(?) to violate our "Once a request has been redirected all future messages referencing that requestId are about the new destination, not the original" invariant on that method.

It's not clear to me what this invariant is, can you explain more? I mean, I see the comment, but what code is making assumptions based on this?

Reply:
> > updated headers
> 
> Why would the headers be updated / differ between these events?

my understanding is that a `Network.responseReceivedExtraInfo`, when it exists, is the source of truth because it comes from the network stack in the browser process, while `Network.responseReceived` comes from the renderer process and so may have imperfect information about the response (e.g. site isolation prevents the renderer process from knowing some details about some cross-origin requests).

> > We handle that with unique requestIds and _findRealRequestAndSetSession, but the extra-info-event ordering would seem(?) to violate our "Once a request has been redirected all future messages referencing that requestId are about the new destination, not the original" invariant on that method.
> 
> It's not clear to me what this invariant is, can you explain more? I mean, I see the comment, but what code is making assumptions based on this?

DT requests all share an ID (e.g. `'1'`) in a chain of redirects, but we change their IDs to e.g. `'1'`, `'1:redirect'`, `'1:redirect:redirect'`, etc to make it easier to keep track of things. To map from an incoming `Network` event with a `requestId` to a particular `NetworkRequest` instance in LH, we assume than any new event has to be for the last request in that chain because we already have all information about any previous redirected requests. That appears to not be the case for the `ExtraInfo` events, e.g. one could be amending the request before the most recent one.

Reply:
I have no extra insight to share really. Matching the extraInfo sequence invariant seems like what we need to do to handle the redirect case.

Reply:
https://chromium-review.googlesource.com/c/chromium/src/+/2898747

Reply:
yeah, that should make things significantly easier. It also seems like some things might break without that after https://chromium-review.googlesource.com/c/chromium/src/+/2856099 lands?